server/clustering_test.go

```diff
@@ -0,0 +1,323 @@
+
+	// Configure first server
+	s1sOpts := getTestDefaultOptsForClustering("a", []string{"b", "c"})
+	s1, err := RunServerWithOpts(s1sOpts, nil)
+	if err != nil {
+		t.Fatalf("Error starting server: %v", err)
+	}
+	defer s1.Shutdown()
+
+	// Configure second server.
+	s2sOpts := getTestDefaultOptsForClustering("b", []string{"a", "c"})
+	s2, err := RunServerWithOpts(s2sOpts, nil)
+	if err != nil {
+		t.Fatalf("Error starting server: %v", err)
+	}
+	defer s2.Shutdown()
+
+	// Configure third server.
+	s3sOpts := getTestDefaultOptsForClustering("c", []string{"a", "b"})
+	s3, err := RunServerWithOpts(s3sOpts, nil)
+	if err != nil {
+		t.Fatalf("Error starting server: %v", err)
+	}
+	defer s3.Shutdown()
+
+	servers := []*StanServer{s1, s2, s3}
+
+	// Create a client connection.
+	sc, err := stan.Connect(clusterName, clientName, stan.PubAckWait(5*time.Second))
+	if err != nil {
+		t.Fatalf("Expected to connect correctly, got err %v", err)
+	}
+	defer sc.Close()
+
+	// Publish a message (this will create the channel and form the Raft group).
+	channel := "foo"
+	if err := sc.Publish(channel, []byte("hello")); err != nil {
+		t.Fatalf("Unexpected error on publish: %v", err)
+	}
+
+	ch := make(chan *stan.Msg, 100)
+	sub, err := sc.Subscribe(channel, func(msg *stan.Msg) {
+		ch <- msg
+	}, stan.DeliverAllAvailable())
+	if err != nil {
+		t.Fatalf("Error subscribing: %v", err)
+	}
+
+	select {
+	case msg := <-ch:
+		assertMsg(t, msg.MsgProto, []byte("hello"), 1)
+	case <-time.After(2 * time.Second):
+		t.Fatal("expected msg")
+	}
+
+	sub.Unsubscribe()
+
+	stopped := []*StanServer{}
+
+	// Take down the leader.
+	leader := getChannelLeader(t, channel, 5*time.Second, servers...)
+	leader.Shutdown()
+	stopped = append(stopped, leader)
+	servers = removeServer(servers, leader)
+
+	// Wait for the new leader to be elected.
+	leader = getChannelLeader(t, channel, 5*time.Second, servers...)
+
+	// Publish some more messages.
+	for i := 0; i < 5; i++ {
+		if err := sc.Publish(channel, []byte(strconv.Itoa(i))); err != nil {
+			t.Fatalf("Unexpected error on publish %d: %v", i, err)
+		}
+	}
+
+	// Read everything back from the channel.
+	sub, err = sc.Subscribe(channel, func(msg *stan.Msg) {
+		ch <- msg
+	}, stan.DeliverAllAvailable())
+	if err != nil {
+		t.Fatalf("Error subscribing: %v", err)
+	}
+	select {
+	case msg := <-ch:
+		assertMsg(t, msg.MsgProto, []byte("hello"), 1)
+	case <-time.After(2 * time.Second):
+		t.Fatal("expected msg")
+	}
+	for i := 0; i < 5; i++ {
+		select {
+		case msg := <-ch:
+			assertMsg(t, msg.MsgProto, []byte(strconv.Itoa(i)), uint64(i+2))
+		case <-time.After(2 * time.Second):
+			t.Fatal("expected msg")
+		}
+	}
+
+	sub.Unsubscribe()
+
+	// Take down the leader.
+	leader.Shutdown()
+	stopped = append(stopped, leader)
+	servers = removeServer(servers, leader)
+
+	// New subscriptions should be rejected since no leader can be established.
+	_, err = sc.Subscribe(channel, func(msg *stan.Msg) {}, stan.DeliverAllAvailable())
```

Comment:
This is the downside of relying on interest-based pub/sub for NATS Streaming. This is *supposed* to timeout because there is no leader, but I'm not sure how we can fast-path this.

Reply:
You can pass the `ConnectWait()` in `stan.Connect()` to reduce the timeout. The timeout is used on Connect and other internal Request/Reply protocols. The danger on reducing it too much is that it may work 100% of cases on local, but flap on Travis (or when running with `-race`)